op-node/rollup/driver/synchronous.go

```diff
@@ -0,0 +1,78 @@
+package driver
+
+import (
+	"context"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+)
+
+// Don't queue up an endless number of events.
+// At some point it's better to drop events and warn something is exploding the number of events.
+const sanityEventLimit = 1000
+
+// SynchronousEvents is a rollup.EventEmitter that a rollup.Deriver can emit events to.
+// The events will be queued up, and can then be executed synchronously by calling the Drain function,
+// which will apply all events to the root Deriver that
+// New events may be queued up while events are being processed by the root rollup.Deriver.
+type SynchronousEvents struct {
+	// The lock is no-op in FP execution, if running in synchronous FP-VM.
+	// This lock ensures that all emitted events are merged together correctly,
+	// if this util is used in a concurrent context.
+	evLock sync.Mutex
+
+	events []rollup.Event
+
+	log log.Logger
+
+	ctx context.Context
+
+	root rollup.Deriver
+}
+
+func NewSynchronousEvents(log log.Logger, ctx context.Context, root rollup.Deriver) *SynchronousEvents {
+	return &SynchronousEvents{
+		log:  log,
+		ctx:  ctx,
+		root: root,
+	}
+}
+
+func (s *SynchronousEvents) Emit(event rollup.Event) {
+	s.evLock.Lock()
+	defer s.evLock.Unlock()
+
+	if s.ctx.Err() != nil {
+		s.log.Warn("Ignoring emitted event during shutdown", "event", event)
+		return
+	}
+
+	// sanity limit, never queue too many events
+	if len(s.events) >= sanityEventLimit {
+		s.log.Error("something is very wrong, queued up too many events! Dropping event", "ev", event)
```

Comment:
```suggestion
		s.log.Error("Something is very wrong, queued up too many events! Dropping event", "ev", event)
```

One day I will just accept that I'll never get log messages to be capitalised correctly, but today is not that day... :)